server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/projection/cassandra/EnqueuedMailsDao.java

```diff
@@ -0,0 +1,156 @@
+import static org.apache.james.queue.rabbitmq.projection.cassandra.EnqueuedMailsTable.HEADER_NAME;
+import static org.apache.james.queue.rabbitmq.projection.cassandra.EnqueuedMailsTable.HEADER_VALUE;
+import static org.apache.james.queue.rabbitmq.projection.cassandra.EnqueuedMailsTable.LAST_UPDATED;
+import static org.apache.james.queue.rabbitmq.projection.cassandra.EnqueuedMailsTable.MAIL_KEY;
+import static org.apache.james.queue.rabbitmq.projection.cassandra.EnqueuedMailsTable.PER_RECIPIENT_SPECIFIC_HEADERS;
+import static org.apache.james.queue.rabbitmq.projection.cassandra.EnqueuedMailsTable.QUEUE_NAME;
+import static org.apache.james.queue.rabbitmq.projection.cassandra.EnqueuedMailsTable.RECIPIENTS;
+import static org.apache.james.queue.rabbitmq.projection.cassandra.EnqueuedMailsTable.REMOTE_ADDR;
+import static org.apache.james.queue.rabbitmq.projection.cassandra.EnqueuedMailsTable.REMOTE_HOST;
+import static org.apache.james.queue.rabbitmq.projection.cassandra.EnqueuedMailsTable.SENDER;
+import static org.apache.james.queue.rabbitmq.projection.cassandra.EnqueuedMailsTable.STATE;
+import static org.apache.james.queue.rabbitmq.projection.cassandra.EnqueuedMailsTable.TABLE_NAME;
+import static org.apache.james.queue.rabbitmq.projection.cassandra.EnqueuedMailsTable.TIME_RANGE_START;
+
+import java.nio.ByteBuffer;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import javax.inject.Inject;
+import javax.mail.internet.AddressException;
+
+import org.apache.james.backends.cassandra.init.configuration.ClusterConfiguration;
+import org.apache.james.backends.cassandra.utils.CassandraAsyncExecutor;
+import org.apache.james.backends.cassandra.utils.CassandraUtils;
+import org.apache.james.core.MailAddress;
+import org.apache.james.queue.rabbitmq.projection.cassandra.model.CassandraMailQueueItem;
+import org.apache.james.util.CompletableFutureUtil;
+import org.apache.mailet.PerRecipientHeaders;
+
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.Row;
+import com.datastax.driver.core.Session;
+import com.datastax.driver.core.UDTValue;
+import com.github.fge.lambdas.Throwing;
+import com.github.steveash.guavate.Guavate;
+
+public class EnqueuedMailsDao {
+
+    private final CassandraAsyncExecutor executor;
+    private final PreparedStatement selectFrom;
+    private final CassandraUtils cassandraUtils;
+    private final Integer bucketSize;
+
+    @Inject
+    public EnqueuedMailsDao(Session session, CassandraUtils cassandraUtils, ClusterConfiguration clusterConfiguration) {
+        this.executor = new CassandraAsyncExecutor(session);
+        this.cassandraUtils = cassandraUtils;
+        this.bucketSize = clusterConfiguration.getHosts().size();
```

Comment:
Nope.

In order to not creqte bugs, this value has to be strictly increasing. (If I have 6 buckets then 5, I might ignore the 6th bucket of the previous slides, thus ignoring some emails, and hurting correctness). You cannot ensure the host count never go down.

More generallly, you need a configuration POJO for slide size and bucket count...